File: src/utils/indicators.ts
export function ema(values: number[], period: number): number {
  if (values.length < period) {
    throw new Error(`Not enough values for EMA(${period})`);
  }
  const alpha = 2 / (period + 1);
  let current = sma(values.slice(0, period));
  for (let i = period; i < values.length; i += 1) {
    current = values[i] * alpha + current * (1 - alpha);
  }
  return current;
}

export function rsi(values: number[], period: number): number {
  if (values.length < period + 1) {
    throw new Error(`Not enough values for RSI(${period})`);
  }

  let gains = 0;
  let losses = 0;
  for (let i = 1; i <= period; i += 1) {
    const delta = values[i] - values[i - 1];
    gains += Math.max(delta, 0);
    losses += Math.max(-delta, 0);
  }

  let avgGain = gains / period;
  let avgLoss = losses / period;

  for (let i = period + 1; i < values.length; i += 1) {
    const delta = values[i] - values[i - 1];
    const gain = Math.max(delta, 0);
    const loss = Math.max(-delta, 0);
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
  }

  if (avgLoss === 0) {
    return 100;
  }
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function sma(values: number[]): number {
  if (values.length === 0) {
    throw new Error("SMA requires at least one value");
  }
  return values.reduce((acc, v) => acc + v, 0) / values.length;
}

export function pctChange(start: number, end: number): number {
  if (start === 0) {
    return 0;
  }
  return ((end - start) / start) * 100;
}

export function atr(
  highs: number[],
  lows: number[],
  closes: number[],
  period: number
): number {
  if (highs.length !== lows.length || highs.length !== closes.length) {
    throw new Error("ATR arrays must have same length");
  }
  if (highs.length < period + 1) {
    throw new Error(`Not enough values for ATR(${period})`);
  }

  const trueRanges: number[] = [];
  for (let i = 1; i < highs.length; i += 1) {
    const tr1 = highs[i] - lows[i];
    const tr2 = Math.abs(highs[i] - closes[i - 1]);
    const tr3 = Math.abs(lows[i] - closes[i - 1]);
    trueRanges.push(Math.max(tr1, tr2, tr3));
  }

  let atrValue = sma(trueRanges.slice(0, period));
  for (let i = period; i < trueRanges.length; i += 1) {
    atrValue = (atrValue * (period - 1) + trueRanges[i]) / period;
  }
  return atrValue;
}

File: python_app/algo_trading_py/utils/indicators.py
from __future__ import annotations

from math import sqrt


def sma(values: list[float]) -> float:
  if not values:
    raise ValueError("sma requires non-empty values")
  return sum(values) / len(values)


def ema(values: list[float], period: int) -> float:
  if period <= 1 or len(values) < period:
    raise ValueError("ema requires len(values) >= period and period > 1")
  k = 2 / (period + 1)
  current = sma(values[:period])
  for v in values[period:]:
    current = v * k + current * (1 - k)
  return current


def rsi(values: list[float], period: int) -> float:
  if period <= 1 or len(values) < period + 1:
    raise ValueError("rsi requires len(values) >= period + 1")
  gains: list[float] = []
  losses: list[float] = []
  for i in range(1, len(values)):
    diff = values[i] - values[i - 1]
    gains.append(max(diff, 0))
    losses.append(max(-diff, 0))
  avg_gain = sma(gains[:period])
  avg_loss = sma(losses[:period])
  for i in range(period, len(gains)):
    avg_gain = (avg_gain * (period - 1) + gains[i]) / period
    avg_loss = (avg_loss * (period - 1) + losses[i]) / period
  if avg_loss == 0:
    return 100.0
  rs = avg_gain / avg_loss
  return 100 - (100 / (1 + rs))


def atr(highs: list[float], lows: list[float], closes: list[float], period: int) -> float:
  if period <= 1 or len(highs) != len(lows) or len(lows) != len(closes) or len(highs) < period + 1:
    raise ValueError("invalid input for atr")
  tr: list[float] = []
  for i in range(1, len(highs)):
    h = highs[i]
    l = lows[i]
    pc = closes[i - 1]
    tr.append(max(h - l, abs(h - pc), abs(l - pc)))
  current = sma(tr[:period])
  for v in tr[period:]:
    current = ((current * (period - 1)) + v) / period
  return current


def pct_change(start: float, end: float) -> float:
  if start == 0:
    raise ValueError("pct_change start cannot be zero")
  return (end - start) / start


def std(values: list[float]) -> float:
  if len(values) < 2:
    return 0.0
  mu = sma(values)
  variance = sum((v - mu) ** 2 for v in values) / len(values)
  return sqrt(variance)
